docs/x509/tutorial.rst

```diff
@@ -0,0 +1,79 @@
+Tutorial
+========
+
+X.509 certificates are used to authenticate clients on servers. The most common
+use case is for web servers using HTTPS.
+
+Creating a Certificate Signing Request (CSR)
+--------------------------------------------
+
+When obtaining a certificate from a certificate authority (CA), the usual
+flow is:
+
+1. You generate a private/public key pair.
+2. You create a request for a certificate, which is signed by your key (to
+   prove that you own that key).
+3. You give your CSR to a CA (but *not* the private key).
+4. The CA validates that you own the resource (e.g. domain) you want a
+   certificate for.
+5. The CA gives you a certificate, signed by them, which identifies your public
+   key, and the resource you are authenticated for.
+6. You configure your server to use that certificate, combined with your
+   private key, to server traffic.
+
+If you want to obtain a certificate from a typical commercial CA, here's how.
+First, you'll need to generate a private key, we'll generate an RSA key (these
+are the most common types of keys on the web right now):
+
+.. code-block:: pycon
+
+    >>> from cryptography.hazmat.backends import default_backend
+    >>> from cryptography.hazmat.primitives import serialization
+    >>> from cryptography.hazmat.primitives.asymmetric import rsa
+    >>> # Generate our key
+    >>> key = rsa.generate_private_key(
+    ...     public_exponent=65537,
+    ...     key_size=2048,
+    ...     backend=default_backend()
+    ... )
+    >>> # Write our key to disk for safe keeping
+    >>> with open("path/to/store/key.pem") as f:
+    ...     f.write(key.private_bytes(
+    ...         encoding=serialization.Encoding.PEM,
+    ...         format=serialization.PrivateFormat.TraditionalOpenSSL,
+    ...         encryption_algorithm=serialization.BestAvailableEncryption(b"passphrase"),
+    ...     ))
+
+If you've already generated a key you can load it with
+:func:`~cryptography.hazmat.primitives.serialization.load_pem_public_key`.
```

Comment:
fixed

On Thu, Jul 16, 2015 at 6:59 AM, Paul Kehrer <notifications@github.com>
wrote:

> In docs/x509/tutorial.rst
> <https://github.com/pyca/cryptography/pull/2154#discussion_r34776210>:
>
> > +    >>> # Generate our key
> > +    >>> key = rsa.generate_private_key(
> > +    ...     public_exponent=65537,
> > +    ...     key_size=2048,
> > +    ...     backend=default_backend()
> > +    ... )
> > +    >>> # Write our key to disk for safe keeping
> > +    >>> with open("path/to/store/key.pem") as f:
> > +    ...     f.write(key.private_bytes(
> > +    ...         encoding=serialization.Encoding.PEM,
> > +    ...         format=serialization.PrivateFormat.TraditionalOpenSSL,
> > +    ...         encryption_algorithm=serialization.BestAvailableEncryption(b"passphrase"),
> > +    ...     ))
> > +
> > +If you've already generated a key you can load it with
> > +:func:`~cryptography.hazmat.primitives.serialization.load_pem_public_key`.
>
> This should be load_pem_private_key?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/cryptography/pull/2154/files#r34776210>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084